doc/sphinxext/missing_references.py

```diff
@@ -0,0 +1,279 @@
+
+
+def get_location(node, app):
+    """
+    Given a docutils node and a sphinx application, return a string
+    representation of the source location of this node.
+
+    Usually, this will be of the form "path/to/file:linenumber". Two
+    special values can be emitted, "<external>" for paths which are
+    not contained in this source tree (e.g. docstrings included from
+    other modules) or "<unknown>", inidcating that the sphinx application
+    cannot locate the original source file (usually because an extension
+    has injected text into the sphinx parsing engine).
+    """
+    (path, line) = get_source_line(node)
+
+    if path:
+
+        # We locate references relative to the parent of the doc
+        # directory, which for matplotlib, will be the root of the
+        # matplotlib repo. When matplotlib is not an editable install
+        # wierd things will happen, but we can't totally recover from
+        # that.
+        basepath = Path(app.srcdir).parent.resolve()
+
+        fullpath = Path(path).resolve()
+
+        try:
+            path = fullpath.relative_to(basepath)
+        except ValueError:
+            # Sometimes docs directly contain e.g. docstrings
+            # from installed modules, and we record those as
+            # <external> so as to be independent of where the
+            # module was installed
+            path = Path("<external>") / fullpath.name
+
+        # Ensure that all reported paths are POSIX so that docs
+        # on windows result in the same warnings in the JSON file.
+        path = path.as_posix()
+
+    else:
+        path = "<unknown>"
+
+    if not line:
+        line = ""
+
+    return f"{path}:{line}"
+
+
+def _warn_unused_missing_references(app):
+    if not app.config.missing_references_warn_unused_ignores:
+        return
+
+    # We can only warn if we are building from a source install
+    # otherwise, we just have to skip this step.
+    basepath = Path(matplotlib.__file__).parent.parent.parent.resolve()
+    srcpath = Path(app.srcdir).parent.resolve()
+
+    if basepath != srcpath:
+        return
+
+    # This is a dictionary of {(domain_type, target): locations}
+    references_ignored = getattr(app.env,
+        'missing_references_ignored_references', {})
+    references_events = getattr(app.env, 'missing_references_events', {})
+
+    # Warn about any reference which is no longer missing.
+    for (domain_type, target), locations in references_ignored.items():
+        missing_reference_locations = references_events.get(
+                (domain_type, target), [])
+
+        # For each ignored reference location, ensure a missing reference
+        # was observed. If it wasn't observed, issue a warning.
+        for ignored_reference_location in locations:
+            if (ignored_reference_location not in
+                    missing_reference_locations):
+                msg = (f"Reference {domain_type} {target} for "
```

Comment:
indent a bit weird

Reply:
Agreed – @QuLogic what were you going for with your earlier comment about this if statement?

Reply:
This indent is correct now; I don't know what @anntzer finds weird.

Reply:
Though I suppose I could point out that the `if` condition _could_ fit on one line.

Reply:
It's the f-string indent which is currently
```python
                msg = (f"Reference {domain_type} {target} for "
                    f"{ignored_reference_location} can be removed"
                    f" from {app.config.missing_references_filename}."
                        "It is no longer a missing reference in the docs.")
```
but should rather be something like
```python
                msg = (f"Reference {domain_type} {target} for "
                       f"{ignored_reference_location} can be removed"
                       f" from {app.config.missing_references_filename}."
                        "It is no longer a missing reference in the docs.")
```